src/probe/probe.c

```diff
@@ -15,42 +22,529 @@
+
+			return -EINVAL;
+		}
+
+		dev = ipc_glb_get_comp_by_id(probe[i].buffer_id);
+		if (!dev) {
+			trace_probe_error("probe_point_add() error: No device with ID %u found.",
+					  probe[i].buffer_id);
+
+			return -EINVAL;
+		}
+
+		if (dev->type != COMP_TYPE_BUFFER) {
+			trace_probe_error("probe_point_add() error: Device ID %u is not a buffer.",
+					  probe[i].buffer_id);
+
+			return -EINVAL;
+		}
+
+		first_free = CONFIG_PROBE_POINTS_MAX;
+
+		for (j = 0; j < CONFIG_PROBE_POINTS_MAX; j++) {
+			if (_probe->probe_points[j].stream_tag ==
+			    PROBE_POINT_INVALID) {
+				if (first_free == CONFIG_PROBE_POINTS_MAX)
+					first_free = j;
+
+				continue;
+			}
+
+			buffer_id = _probe->probe_points[j].buffer_id;
+			if (buffer_id == probe[i].buffer_id) {
+				if (_probe->probe_points[j].purpose ==
+				    probe[i].purpose) {
+					trace_probe_error("probe_point_add() error: Probe already attached to buffer %u with purpose %u",
+							  buffer_id,
+							  probe[i].purpose);
+
+					return -EINVAL;
+				}
+			}
+		}
+
+		if (first_free == CONFIG_PROBE_POINTS_MAX) {
+			trace_probe_error("probe_point_add() error: Maximum number of probe points connected aleady: " META_QUOTE(CONFIG_PROBE_POINTS_MAX));
+
+			return -EINVAL;
+		}
+
+		/* if connecting injection probe, check for associated DMA */
+		if (probe[i].purpose == PROBE_PURPOSE_INJECTION) {
+			dma_found = 0;
+
+			for (j = 0; j < CONFIG_PROBE_DMA_MAX; j++) {
+				if (_probe->inject_dma[j].stream_tag !=
+				    PROBE_DMA_INVALID &&
+				    _probe->inject_dma[j].stream_tag ==
+				    probe[i].stream_tag) {
+					dma_found = 1;
+					break;
+				}
+			}
+
+			if (!dma_found) {
+				trace_probe_error("probe_point_add() error: No DMA with stream tag %u found for injection.",
+						  probe[i].stream_tag);
+
+				return -EINVAL;
+			}
+		}
+
+		_probe->probe_points[first_free].buffer_id = probe[i].buffer_id;
+		_probe->probe_points[first_free].purpose = probe[i].purpose;
+		_probe->probe_points[first_free].stream_tag =
+			probe[i].stream_tag;
+
+		/* TODO: Hook up callbacks to buffer (dev->cb) and DMA */
+	}
+
 	return 0;
 }
 
 int probe_point_info(struct sof_ipc_probe_info_params *data, uint32_t max_size)
 {
-	return 0;
+	uint32_t i = 0;
+	uint32_t j = 0;
+	uint32_t *size = &data->rhdr.hdr.size;
+
+	tracev_probe("probe_point_info()");
+
+	if (!_probe) {
+		trace_probe_error("probe_point_info() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	*size = sizeof(*data);
+
+	while (*size + sizeof(struct probe_point) < max_size &&
+	       i < CONFIG_PROBE_POINTS_MAX) {
+		if (_probe->probe_points[i].stream_tag != PROBE_POINT_INVALID) {
+			data->probe_point[j].buffer_id =
+				_probe->probe_points[i].buffer_id;
+			data->probe_point[j].purpose =
+				_probe->probe_points[i].purpose;
+			data->probe_point[j].stream_tag =
+				_probe->probe_points[i].stream_tag;
+			j++;
+			*size += sizeof(struct probe_point);
+		}
+
+		i++;
+	}
+
+	return 1;
```

Comment:
what does 1 mean ? 0 means success.

Reply:
1 means "reply created and copied by cmd()" as described in ipc_cmd()
Already discussed in PR #2205